src/components/percentage/percentage.test.jsx

```diff
@@ -0,0 +1,261 @@
+    });
+  });
+
+  describe('onChange API', () => {
+    it('sets the onchange handler', () => {
+      const onChangeSpy = jest.fn();
+      render(<Percentage {...requiredProps} onChange={onChangeSpy} />);
+      fireEvent.change(screen.getByLabelText('the label'), { target: { value: '100' } });
+      expect(onChangeSpy.mock.calls.length).toEqual(1);
+    });
+  });
+
+  describe('onFocus API', () => {
+    it('sets the onFocus handler', () => {
+      const onFocusSpy = jest.fn();
+      render(<Percentage {...requiredProps} onFocus={onFocusSpy} />);
+      fireEvent.focus(screen.getByLabelText('the label'));
+      expect(onFocusSpy.mock.calls.length).toEqual(1);
+    });
+  });
+
+  describe('onInput API', () => {
+    it('sets the onInput handler', () => {
+      const onInputSpy = jest.fn();
+      render(<Percentage {...requiredProps} onInput={onInputSpy} />);
+      fireEvent.input(screen.getByLabelText('the label'));
+      expect(onInputSpy.mock.calls.length).toEqual(1);
+    });
+  });
+
+  describe('onKeyDown API', () => {
+    it('sets the onKeyDown handler', () => {
+      const onKeyDownSpy = jest.fn();
+      render(<Percentage {...requiredProps} onKeyDown={onKeyDownSpy} />);
+      fireEvent.keyDown(screen.getByLabelText('the label'));
+      expect(onKeyDownSpy.mock.calls.length).toEqual(1);
+    });
+  });
+
+  describe('placeholder API', () => {
+    it('sets the placeholder attribute', () => {
+      render(<Percentage {...requiredProps} placeholder="test-placeholder" />);
+      expect(screen.getByLabelText('the label')).toHaveAttribute('placeholder', 'test-placeholder');
+    });
+  });
+
+  describe('readOnly API', () => {
+    it('sets the readOnly attribute', () => {
+      render(<Percentage {...requiredProps} readOnly />);
+      expect(screen.getByLabelText('the label')).toHaveAttribute('readonly');
+    });
+  });
+
+  describe('ref API', () => {
+    it('can be set', () => {
+      const ref = React.createRef();
+      render(<Percentage {...requiredProps} ref={ref} value={100} />);
+      expect(ref.current.value).toEqual(100);
+    });
+  });
+
+  describe('required API', () => {
+    it('can be set', () => {
+      render(<Percentage {...requiredProps} required={true} />);
+      expect(screen.getByLabelText('the label')).toBeRequired();
+      expect(screen.getByLabelText('the label')).toBeInvalid();
+    });
+
+    it('can be unset', () => {
+      render(<Percentage {...requiredProps} required={false} />);
+      expect(screen.getByLabelText('the label')).not.toBeRequired();
+      expect(screen.getByLabelText('the label')).toBeValid();
+    });
+
+    it('is invalid on mount but does not have an error message', () => {
+      render(<Percentage {...requiredProps} required={true} />);
+      expect(screen.getByLabelText('the label')).toBeInvalid();
+      expect(screen.getByLabelText('the label')).toHaveDescription('');
+    });
+
+    it('is invalid after typing', async () => {
```

Comment:
This is no longer part of the ref spec. It should not have an error message after typing. Only after blurring.